File: sponge-examples/examples/script/rb/knowledge_base_auto_enable.rb
# Sponge Knowledge base
# Auto-enable

java_import java.util.concurrent.atomic.AtomicInteger

def onInit
    # Variables for assertions only
    $EPS.setVariable("counter", AtomicInteger.new(0))
end

class AutoAction < Action
    def run(args)
        self.logger.debug("Running")
        $EPS.getVariable("counter").incrementAndGet()
        return args
    end
end

class AutoFilter < Filter
    def configure
        self.eventName = "e1"
    end
    def accepts(event)
        self.logger.debug("Received event: {}", event.name)
        $EPS.getVariable("counter").incrementAndGet()
        return true
    end
end

class AutoTrigger < Trigger
    def configure
        self.eventName = "e1"
    end
    def run(event)
        self.logger.debug("Received event: {}", event.name)
        $EPS.getVariable("counter").incrementAndGet()
    end
end

class AutoRule < Rule
    def configure
        self.events = ["e1", "e2"]
    end
    def run(event)
        self.logger.debug("Running for sequence: {}", self.eventSequence)
        $EPS.getVariable("counter").incrementAndGet()
    end
end

class AutoAggregator < Aggregator
    def configure
        self.eventNames = ["e1", "e2"]
    end
    def acceptsAsFirst(event)
        return event.name == "e1"
    end
    def onEvent(event)
        self.logger.debug("Received event: {}", event.name)
        if event.name == "e2"
            $EPS.getVariable("counter").incrementAndGet()
            self.finish()
        end
    end
end

def onStartup
    $EPS.callAction("AutoAction")
    $EPS.event("e1").send()
    $EPS.event("e2").send()
end

File: sponge-integration-tests/src/test/java/org/openksavi/sponge/integration/tests/java/examples/KnowledgeBaseAutoEnable.java
/*
 * Copyright 2016-2019 The Sponge authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.openksavi.sponge.integration.tests.java.examples;

import java.util.concurrent.atomic.AtomicInteger;

import org.openksavi.sponge.event.Event;
import org.openksavi.sponge.java.JAction;
import org.openksavi.sponge.java.JCorrelator;
import org.openksavi.sponge.java.JFilter;
import org.openksavi.sponge.java.JKnowledgeBase;
import org.openksavi.sponge.java.JRule;
import org.openksavi.sponge.java.JTrigger;

/**
 * Sponge Knowledge Base. Auto-enable.
 */
public class KnowledgeBaseAutoEnable extends JKnowledgeBase {

    @Override
    public void onInit() {
        // Variables for assertions only
        getSponge().setVariable("counter", new AtomicInteger(0));
    }

    public static class AutoAction extends JAction {

        public Object onCall() {
            getLogger().debug("Running");
            getSponge().getVariable(AtomicInteger.class, "counter").incrementAndGet();
            return null;
        }
    }

    public static class AutoFilter extends JFilter {

        @Override
        public void onConfigure() {
            withEvent("e1");
        }

        @Override
        public boolean onAccept(Event event) {
            getLogger().debug("Received event: {}", event.getName());
            getSponge().getVariable(AtomicInteger.class, "counter").incrementAndGet();
            return true;
        }
    }

    public static class AutoTrigger extends JTrigger {

        @Override
        public void onConfigure() {
            withEvent("e1");
        }

        @Override
        public void onRun(Event event) {
            getLogger().debug("Received event: {}", event.getName());
            getSponge().getVariable(AtomicInteger.class, "counter").incrementAndGet();
        }
    }

    public static class AutoRule extends JRule {

        @Override
        public void onConfigure() {
            withEvents("e1", "e2");
        }

        @Override
        public void onRun(Event event) {
            getLogger().debug("Running for sequence: {}", getEventSequence());
            getSponge().getVariable(AtomicInteger.class, "counter").incrementAndGet();
        }
    }

    public static class AutoCorrelator extends JCorrelator {

        @Override
        public void onConfigure() {
            withEvents("e1", "e2");
        }

        @Override
        public boolean onAcceptAsFirst(Event event) {
            return event.getName().equals("e1");
        }

        @Override
        public void onEvent(Event event) {
            getLogger().debug("Received event: {}", event.getName());
            if (event.getName().equals("e2")) {
                getSponge().getVariable(AtomicInteger.class, "counter").incrementAndGet();
                finish();
            }
        }
    }

    @Override
    public void onStartup() {
        getSponge().call("AutoAction");
        getSponge().event("e1").send();
        getSponge().event("e2").send();
    }
}
